lib/path.nix

```diff
@@ -0,0 +1,174 @@
+      throw "${errorPrefix}: The string is an absolute path because it starts with `/`"
+    else true;
+
+  # Splits and normalises a subpath string into its components.
+  # Errors for ".." components and doesn't include "." components
+  splitSubpath = path: errorPrefix:
+    let
+      # Split the string into its parts using regex for efficiency. This regex
+      # matches patterns like "/", "/./", "/././", with arbitrarily many "/"s
+      # together. These are the main special cases:
+      # - Leading "./" gets split into a leading "." part
+      # - Trailing "/." or "/" get split into a trailing "." or ""
+      #   part respectively
+      #
+      # These are the only cases where "." and "" parts can occur
+      parts = split "/+(\\./+)*" path;
+
+      # `split` creates a list of 2 * k + 1 elements, containing the k +
+      # 1 parts, interleaved with k matches where k is the number of
+      # (non-overlapping) matches. This calculation here gets the number of parts
+      # back from the list length
+      # floor( (2 * k + 1) / 2 ) + 1 == floor( k + 1/2 ) + 1 == k + 1
+      partCount = length parts / 2 + 1;
+
+      # To assemble the final list of components we want to:
+      # - Skip a potential leading ".", normalising "./foo" to "foo"
+      # - Skip a potential trailing "." or "", normalising "foo/" and "foo/." to
+      #   "foo"
+      skipStart = if head parts == "." then 1 else 0;
+      skipEnd = if last parts == "." || last parts == "" then 1 else 0;
+
+      # We can now know the length of the result by removing the number of
+      # skipped parts from the total number
+      componentCount = partCount - skipEnd - skipStart;
+
+    in
+      # Special case of a single "." path component. Such a case leaves a
+      # componentCount of -1 due to the skipStart/skipEnd not verifying that
+      # they don't refer to the same character
+      if path == "." then []
+
+      # And we can use this to generate the result list directly. Doing it this
+      # way over a combination of `filter`, `init` and `tail` makes it more
+      # efficient, because we don't allocate any intermediate lists
+      else genList (index:
+        let
+          # To get to the element we need to add the number of parts we skip and
+          # multiply by two due to the interleaved layout of `parts`
+          value = elemAt parts ((skipStart + index) * 2);
+        in
+
+        # We don't support ".." components, see ./path-design.md
+        if value == ".." then
+          throw "${errorPrefix}: Path string contains contains a `..` component, which is not supported"
+        # Otherwise just return the part unchanged
+        else
+          value
+      ) componentCount;
+
+  # joins subpath components together
+  joinSubpath = components:
+    "./" +
+    # An empty string is not a valid subpath, so we need to return a `.` when we have no components
+    (if components == [] then "."
+    else concatStringsSep "/" components);
```

Comment:
Eh I think it's not a problem to have extra comments. I introduced this comment because originally I implemented it as just `concatStringsSep "/"` but then I noticed how `.` doesn't work, so I added this special case. It's not obvious that "" is not a valid relative path.

I'm adding your extra comment though